mls-rs-uniffi/src/config.rs

```diff
@@ -1,43 +1,79 @@
 
 use mls_rs::{
     client_builder::{self, WithGroupStateStorage},
     identity::basic,
+    storage_provider::in_memory::InMemoryGroupStateStorage,
 };
-use mls_rs_core::error::IntoAnyError;
 use mls_rs_crypto_openssl::OpensslCryptoProvider;
 
-use self::group_state::{GroupStateStorage, GroupStateStorageWrapper};
+use self::group_state::{GroupStateStorage, GroupStateStorageAdapter};
+use crate::Error;
 
 pub mod group_state;
 
-#[derive(Debug, thiserror::Error, uniffi::Error)]
-#[uniffi(flat_error)]
-#[non_exhaustive]
-pub enum FFICallbackError {
-    #[error("data preparation error")]
-    DataPreparationError {
-        #[from]
-        inner: mls_rs_core::mls_rs_codec::Error,
-    },
-    #[error("unexpected callback error")]
-    UnexpectedCallbackError {
-        #[from]
-        inner: uniffi::UnexpectedUniFFICallbackError,
-    },
+#[derive(Debug, Clone)]
+pub(crate) struct ClientGroupStorage(Arc<dyn GroupStateStorage>);
+
+impl From<Arc<dyn GroupStateStorage>> for ClientGroupStorage {
+    fn from(value: Arc<dyn GroupStateStorage>) -> Self {
+        Self(value)
+    }
 }
 
-impl IntoAnyError for FFICallbackError {}
+#[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+#[cfg_attr(mls_build_async, maybe_async::must_be_async)]
+impl mls_rs_core::group::GroupStateStorage for ClientGroupStorage {
+    type Error = Error;
+
+    async fn state(&self, group_id: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
+        self.0.state(group_id.to_vec())
+    }
+
+    async fn epoch(&self, group_id: &[u8], epoch_id: u64) -> Result<Option<Vec<u8>>, Self::Error> {
+        self.0.epoch(group_id.to_vec(), epoch_id)
+    }
+
+    async fn write(
+        &mut self,
+        state: mls_rs_core::group::GroupState,
+        inserts: Vec<mls_rs_core::group::EpochRecord>,
+        updates: Vec<mls_rs_core::group::EpochRecord>,
+    ) -> Result<(), Self::Error> {
+        self.0.write(
+            state.into(),
+            inserts.into_iter().map(Into::into).collect(),
+            updates.into_iter().map(Into::into).collect(),
+        )
+    }
+
+    async fn max_epoch_id(&self, group_id: &[u8]) -> Result<Option<u64>, Self::Error> {
+        self.0.max_epoch_id(group_id.to_vec())
+    }
+}
 
 pub type UniFFIConfig = client_builder::WithIdentityProvider<
     basic::BasicIdentityProvider,
     client_builder::WithCryptoProvider<
         OpensslCryptoProvider,
-        WithGroupStateStorage<GroupStateStorageWrapper, client_builder::BaseConfig>,
+        WithGroupStateStorage<ClientGroupStorage, client_builder::BaseConfig>,
     >,
 >;
 
 #[derive(Debug, Clone, uniffi::Record)]
 pub struct ClientConfig {
     pub group_state_storage: Arc<dyn GroupStateStorage>,
 }
+
+// TODO(mgeisler): turn into an associated function when UniFFI
+// supports them: https://github.com/mozilla/uniffi-rs/issues/1074.
+/// Create a client config with an in-memory group state storage.
+#[uniffi::export]
+pub fn client_config_default() -> ClientConfig {
```

Comment:
Maybe I should add that I'm not attached to the name here: it could also be something like `with_in_memory_storage` if you like.

I would also like to have an easy way to instantiate the Sqlite storage, similar to [`ClientBuilder::new_sqlite`](https://docs.rs/mls-rs/latest/mls_rs/client_builder/struct.ClientBuilder.html#method.new_sqlite). Perhaps it could be even simpler and just take a path name to the database and then configure all three storages to use a Sqlite database under that path.

Reply:
I like the SQLite idea, we could make them provide a path and a key -- let's make it encrypted by default?

Reply:
That sounds nice indeed — I haven't looked at the Sqlite implementation yet, but does it support on-disk encryption directly? Or do you suggest that we encrypt the file on disk somehow?